README.md
# luch-request v4

`luch-request v4` 是面向传统 uni-app 的 TypeScript-first 请求库。

当前版本处于 `4.0.0-alpha` 阶段，已通过 npm `alpha` dist-tag 发布。

## 当前能力

- `request`、常用 HTTP 方法、`upload`、`download`
- 实例默认配置与单次请求配置
- request/response interceptor
- 统一的 `LuchRequestError`
- Promise 上的 `abort()`、`task` 和 `onTask()`
- 通过 `signal` 传播取消，并提供跨平台 `createCancelSource()`
- `nativeOptions` 动态透传平台新增参数，未知响应字段保留

## 安装

```sh
npm install luch-request@alpha
```

`alpha` 指向当前 v4 预发布版本；不带 `@alpha` 会使用 npm `latest` 通道。

v4 只提供 ESM：

```ts
import {
  CancellationMode,
  createCancelSource,
  createLuchRequest,
  isLuchRequestError,
  LuchRequestError
} from 'luch-request'
```

构建产物使用 ES2017 语法，不提供 CommonJS、UMD、ES5 legacy bundle 或
运行时 polyfill。首个稳定版本计划支持传统 uni-app，App iOS 的正式支持
下限为 iOS 13；不支持 uni-app x / UTS。完整范围和验证要求见
[`COMPATIBILITY.md`](./COMPATIBILITY.md)。

## 快速开始

```ts
import {
  CancellationMode,
  createCancelSource,
  createLuchRequest,
  isLuchRequestError,
  LuchRequestError
} from 'luch-request'

const http = createLuchRequest({
  baseURL: 'https://api.example.com',
  timeout: 10_000
})

try {
  const response = await http.get<{ id: number }>('/users/1')
  console.log(response.data.id)
} catch (error) {
  if (isLuchRequestError(error)) {
    console.log(error.code, error.raw, error.cancelMode)

    if (error.code === LuchRequestError.ERR_CANCELED) {
      console.log('请求已取消')
    }
  }
}
```

## 创建实例

使用 `createLuchRequest(defaults)` 创建独立实例。不同实例的默认配置和
interceptor 不共享：

```ts
const http = createLuchRequest({
  baseURL: 'https://api.example.com',
  method: 'POST',
  timeout: 10_000,
  header: {
    Accept: 'application/json'
  },
  validateStatus: (status) => status >= 200 && status < 300
})
```

### 内置默认行为

`createLuchRequest()` 不传配置时，luch-request 使用以下默认行为：

| 配置 | 用户未传时的行为 |
| --- | --- |
| `baseURL` | 按空字符串处理，直接使用请求的 `url` |
| `method` | `request()` 使用实例默认 method，未配置时为 `GET`；快捷方法使用各自固定 method |
| `header` | 不主动添加请求头 |
| `params` | 不追加查询参数 |
| `paramsSerializer` | 使用内置序列化规则，详见“查询参数” |
| `validateStatus` | 只接受 `200` 至 `299` |
| `luchOptions.jsonParsing` | 对 upload 使用 `auto` 解析；其他 operation 不解析 |
| `luchOptions.isNativeAbortError` | 使用内置窄规则尽力识别原生 Task 的 abort 失败 |
| `signal` | 不监听取消信号 |
| `luchMeta`、`nativeOptions` | 不设置 |
| `timeout` 及其他 uni 原生配置 | 不设置，由目标平台采用自身默认值 |

`method` 和 `validateStatus` 会在 request interceptor 前补齐。普通 request
的 `config.method` 是大写有效值；upload 和 download 没有 method，但同样
包含实际使用的 `validateStatus`。其他默认行为不会全部写入 config，例如用户
未传 `baseURL` 时仍为 `undefined`，只在生成 `fullURL` 时按空字符串处理。
JSON 解析的 `mode` 未传时按 `auto` 处理，`include` 未传时只匹配 upload。

### 配置合并策略

配置优先级固定为“内置行为默认值 → 实例默认配置 → 单次请求配置”，即同一
配置项由靠后的有效值覆盖。不同配置项按以下策略处理：

| 配置 | 合并策略 |
| --- | --- |
| `header` | 按大小写不敏感的 header 名合并；单次值覆盖实例值，`null/undefined` 删除继承值 |
| `luchMeta` | 第一层浅合并；同名嵌套值由单次请求整体替换 |
| `luchOptions` | 按库功能定义策略；`jsonParsing` 对象按字段合并，`isNativeAbortError` 单次函数整体替换实例函数 |
| `nativeOptions` | request 实例默认值与单次 request 第一层浅合并；upload/download 只使用各自单次配置 |
| `method` | 只作为普通 request 的默认值；单次 request 或快捷方法覆盖实例值，upload/download 不继承 |
| `baseURL`、`params`、`paramsSerializer`、`validateStatus`、`signal`、`timeout` | 单次请求整体覆盖实例值 |
| `data`、`formData`、`files` 及其他已声明原生字段 | 单次请求整体覆盖，不与实例值递归合并；采用的普通结构会生成请求私有副本 |
| 普通对象和数组 | 递归复制普通对象及数组中的普通结构，防止 interceptor 污染 defaults、调用方输入或其他请求 |
| `signal`、函数、`Date`、`ArrayBuffer`、类实例和平台对象 | 保留引用身份，不尝试通用克隆 |

例如：

```ts
const http = createLuchRequest({
  params: {
    language: 'zh',
    pageSize: 20
  },
  luchMeta: {
    source: 'app',
    auth: {
      required: true,
      role: 'user'
    }
  },
  nativeOptions: {
    platformOption: {
      cache: true
    }
  }
})

await http.get('/users', {
  params: {
    page: 2
  },
  luchMeta: {
    auth: {
      role: 'admin'
    }
  },
  nativeOptions: {
    platformOption: {
      trace: true
    }
  }
})
```

合并后：

- `params` 是 `{ page: 2 }`，不会与实例 `params` 合并。
- `luchMeta` 保留第一层的 `source`，但 `auth` 整体替换为
  `{ role: 'admin' }`。
- `nativeOptions.platformOption` 整体替换为 `{ trace: true }`。

`header` 还有一项 upload 特例：本次 upload 未显式设置 `Content-Type` 时，
不会继承实例默认值中的该请求头，让平台生成 multipart 类型及 boundary；
其他默认 header 正常继承。

局部 header 使用 `null` 或 `undefined` 可以删除实例继承值；空值会在派发前
再次过滤，不传给底层 uni API：

```ts
await http.get('/users', {
  header: {
    Authorization: undefined
  }
})
```

`luchOptions.jsonParsing` 对象按字段合并，其中 `include` 数组整体替换；
单次请求设置为 `false` 时明确关闭解析。

合并过程不会写入传入的默认配置或单次请求配置。普通对象和数组会生成实例或
请求私有的结构副本，因此 request interceptor 原地修改普通嵌套配置时，不会
污染其他实例、实例 defaults、调用方输入或其他并发请求。该行为不是对所有值
执行通用深克隆：`signal`、函数、`Date`、`ArrayBuffer`、class 实例和平台对象
仍保留引用身份，interceptor 不应修改这些不透明对象的内部状态。

## 请求 API

### 完整配置

```ts
interface User {
  id: number
  name: string
}

interface LoginBody {
  username: string
  password: string
}

interface LoginParams {
  source: string
}

const response = await http.request<
  User,
  LoginBody,
  LoginParams
>({
  url: '/user/login',
  method: 'POST',
  data: {
    username: 'demo',
    password: 'secret'
  },
  params: {
    source: 'uni-app'
  }
})
```

`request<TResponse, TData, TParams, TNativeOptions>()` 的四个泛型依次
表示响应数据、请求体、查询参数和本次请求的 `nativeOptions`。

### 快捷方法

```ts
http.get<TResponse, TParams, TNativeOptions>(url, config)
http.delete<TResponse, TParams, TNativeOptions>(url, config)
http.head<TResponse, TParams, TNativeOptions>(url, config)
http.options<TResponse, TParams, TNativeOptions>(url, config)

http.post<TResponse, TData, TParams, TNativeOptions>(
  url,
  data,
  config
)
http.put<TResponse, TData, TParams, TNativeOptions>(
  url,
  data,
  config
)
http.patch<TResponse, TData, TParams, TNativeOptions>(
  url,
  data,
  config
)
```

示例：

```ts
interface User {
  id: number
  name: string
}

interface CreateUser {
  name: string
}

interface UserListParams {
  page: number
  pageSize: number
  tags?: string[]
}

const list = await http.get<User[], UserListParams>('/users', {
  params: {
    page: 1,
    pageSize: 20,
    tags: ['new', 'active']
  }
})

const created = await http.post<User, CreateUser>(
  '/users',
  {
    name: 'Ada'
  }
)
```

### 常用配置

| 配置 | 说明 |
| --- | --- |
| `baseURL` | 相对 `url` 使用的基础地址；绝对地址不会与它组合 |
| `header` | 发送给平台的请求头 |
| `params` | 追加到 URL 的查询参数 |
| `paramsSerializer` | 自定义查询参数序列化 |
| `luchMeta` | 业务和 interceptor 元数据，不发送给 uni API |
| `validateStatus` | 判断 HTTP 状态是否成功，默认接受 2xx |
| `signal` | 可选的结构化取消信号 |
| `timeout` | 请求超时时间，实际能力由平台决定 |
| `nativeOptions` | 未建模、平台专有或与库配置重名的原生参数 |

`success`、`fail`、`complete` 被内部 callback adapter 保留，不能在配置中覆盖。
`fullURL` 由库内部生成，也不能作为请求配置传入。

### 查询参数

普通 request 的 `TParams` 用于约束输入查询参数，并传播到
`paramsSerializer`：

```ts
interface UserListParams {
  page: number
  pageSize: number
  keyword?: string
}

const response = await http.get<User[], UserListParams>(
  '/users',
  {
    params: {
      page: 1,
      pageSize: 20
    },
    paramsSerializer: (params) => (
      `page=${params.page}&pageSize=${params.pageSize}`
    )
  }
)
```

显式提供 `TParams` 且传入 `params` 时，对象内的参数名、必填字段和值类型
由 TypeScript 检查；低层 client 不强制整个 `params` 必须存在。未显式提供
`TParams` 时，
`params` 使用宽泛的 `object`，保持只指定响应类型时传入普通业务 interface
的便利：

```ts
const params: UserListParams = {
  page: 1,
  pageSize: 20
}

await http.get<User[]>('/users', {
  params
})
```

实例 defaults 和 request interceptor 都可以替换最终 params，因此
`response.config.params` 只承诺为 `object | undefined`。需要在响应阶段关联
业务参数时，应从业务闭包或 `luchMeta` 读取，不依赖无法由运行时保证的泛型。

默认序列化规则：

- 忽略 `undefined`。
- 数组转换为 `key[]=value1&key[]=value2`。
- `Date` 转换为 ISO 字符串。
- 对象使用 `JSON.stringify`。
- URL 中已有查询参数时继续追加；fragment 不会被带入请求。

需要不同格式时：

```ts
interface UserIdsParams {
  ids: number[]
}

await http.get<unknown, UserIdsParams>('/users', {
  params: {
    ids: [1, 2, 3]
  },
  paramsSerializer: (params) => {
    return `ids=${params.ids.join(',')}`
  }
})
```

## TypeScript 与平台新增参数

已知字段在顶层提供直接类型提示。未声明的顶层字段不会传给 uni API；
uni-app 或具体平台新增字段统一放入 `nativeOptions`：

```ts
import type { QueryParams } from 'luch-request'

await http.get<User, QueryParams, {
  enableProfile?: boolean
  futurePlatformOption?: string
}>('/users/1', {
  nativeOptions: {
    enableProfile: true,
    futurePlatformOption: 'enabled'
  }
})
```

额外泛型只描述本次请求。项目多处使用同一批平台参数时，可通过 module
augmentation 一次性补充全局类型：

```ts
// types/luch-request.d.ts
import 'luch-request'

declare module 'luch-request' {
  interface LuchRequestNativeOptions {
    enableProfile?: boolean
    futurePlatformOption?: string
  }

  interface LuchUploadNativeOptions {
    enableBackgroundUpload?: boolean
  }

  interface LuchDownloadNativeOptions {
    useDownloadCache?: boolean
  }
}

export {}
```

之后无需重复泛型：

```ts
await http.get<User>('/users/1', {
  nativeOptions: {
    enableProfile: true
  }
})
```

`nativeOptions` 在派发前做一层展开，不作为嵌套字段发送；通常会覆盖同名的
已声明原生字段。`url`、`fullURL`、`success`、`fail`、`complete` 属于内部
保护字段，即使通过 JavaScript 动态传入也会被忽略。透传不代表所有平台都
支持该参数，调用方仍需根据目标平台文档和基础库版本进行能力判断。

实例 defaults 的类型是 request defaults，其中的 request 专属原生字段和
`nativeOptions` 不会进入 upload/download。文件操作只继承 `baseURL`、
`header`、`params`、`luchMeta`、`validateStatus`、`signal`、`timeout`、
`luchOptions` 等公共配置；其平台扩展应写在本次 upload/download 的
`nativeOptions` 中。已知属于其他 operation 的字段会在派发边界被过滤。

## 响应结构

普通请求返回原生响应字段，并增加请求上下文：

```ts
const response = await http.get<User>('/users/1')

response.data
response.statusCode
response.header
response.config
response.task
response.raw
```

- `data`：用户泛型指定的响应数据。
- `statusCode`：归一化后的数字状态码。
- `config.url`：request interceptor 最终返回的逻辑地址，保留调用方写法。
- `config.fullURL`：在 request interceptor 完成后生成，包含 `baseURL` 和
  `params` 的实际请求地址。
- `config.params`：interceptor 和 defaults 处理后的最终查询参数，类型为宽
  `object | undefined`。
- `task`：平台提供时对应原生 Task。
- `raw`：完整原始平台响应。
- 平台新增响应字段保留在响应顶层，不通过字段白名单重建。

默认只有 2xx 响应 resolve。可以按业务调整：

```ts
const http = createLuchRequest({
  validateStatus: (status) => status >= 200 && status < 400
})
```

## JSON 响应解析

由于 `uni.uploadFile` 固定返回字符串 `data`，luch-request 默认只对 upload
使用 `auto` 模式尝试 `JSON.parse`。普通 request 和 download 不额外解析：

```ts
import { createLuchRequest } from 'luch-request'

const http = createLuchRequest()

interface UploadResult {
  code: number
  data: {
    id: number
  }
}

const response = await http.upload<UploadResult>({
  url: '/upload',
  filePath,
  name: 'file'
})

if (typeof response.data === 'string') {
  console.log('服务端返回的不是有效 JSON', response.data)
} else {
  console.log(response.data.code)
}
```

- 只在 `response.data` 是字符串且当前 operation 位于 `include` 时解析。
- `auto` 为默认模式；解析失败时保留原字符串。
- `strict` 模式解析失败时抛出 `ERR_BAD_RESPONSE`。
- 解析只改变外层 `response.data`，`response.raw.data` 保留原始字符串。
- 未提供 `include` 时只匹配 `LuchOperation.UPLOAD`。
- `upload<T>()` 的 `response.data` 类型为 `T | string`，调用方需要缩小类型。

实例或单次配置可以启用严格模式，也可以明确关闭解析：

```ts
await http.upload({
  url: '/upload',
  filePath,
  name: 'file',
  luchOptions: {
    jsonParsing: {
      mode: 'strict'
    }
  }
})

await http.upload({
  url: '/upload',
  filePath,
  name: 'file',
  luchOptions: {
    jsonParsing: false
  }
})
```

## Interceptor

request 和 response interceptor 均按注册顺序 FIFO 执行，支持同步或异步
返回。普通请求、上传和下载共用实例上的两组 interceptor。`use()` 返回稳定
ID，可用 `eject()` 移除，也可用 `clear()` 清空。公共入口只导出
`InterceptorManager` interface，不导出内部实现类、handler 存储结构或遍历
方法。

handler 的第二个参数提供只读 operation，不需要通过字段猜测请求类型：

```ts
http.interceptors.request.use((config, context) => {
  if (context.operation === LuchOperation.UPLOAD) {
    // upload 专属逻辑
  }

  return config
})
```

`LuchOperation.REQUEST`、`UPLOAD`、`DOWNLOAD` 对应三类公开调用。operation
不写入 config、response 或 error，也不使用 `UPLOAD`/`DOWNLOAD` 伪 method。

进入 request interceptor 前，luch-request 已补齐会影响请求行为的默认值：

- `context.operation` 为 `REQUEST` 时，`config.method` 是大写有效 method。
- `UPLOAD` 或 `DOWNLOAD` 时，config 不包含顶层 `method`；即使实例设置了
  默认 method，也不会传给文件 API。
- 三种 operation 的 `config.validateStatus` 都是实际使用的函数。
- interceptor 可以修改这些值；返回后会再次补默认值、校验并归一化。

### Request interceptor

```ts
const requestInterceptorId = http.interceptors.request.use(
  async (config) => {
    const token = await getAccessToken()

    return {
      ...config,
      header: {
        ...config.header,
        Authorization: `Bearer ${token}`
      }
    }
  },
  (error) => {
    throw error
  }
)

http.interceptors.request.eject(requestInterceptorId)
```

request interceptor 可以读取 `config.luchMeta`：

```ts
await http.get('/users/1', {
  luchMeta: {
    requiresAuth: true,
    traceName: 'user-detail'
  }
})
```

项目可以通过 module augmentation 为 `luchMeta` 补充业务类型：

```ts
declare module 'luch-request' {
  interface LuchMeta {
    requiresAuth?: boolean
    traceName?: string
  }
}
```

此阶段只有逻辑 `config.url`，没有 `config.fullURL`，也不能设置它。
`fullURL` 会在全部 request interceptor 完成后生成，并出现在响应及已进入
派发阶段的错误 `config` 中。

### 在 interceptor 中终止请求

在 request interceptor 中抛出错误，请求不会进入对应的底层 uni API
（普通请求即 `uni.request`）：

```ts
http.interceptors.request.use((config) => {
  if (config.luchMeta?.requiresAuth === true && !hasLogin()) {
    throw new LuchRequestError(
      '用户未登录，终止请求',
      LuchRequestError.ERR_CANCELED
    )
  }

  return config
})
```

请求管线会为该错误补充当前 `config`。因为这里只是通过异常终止流程，没有
执行 `abort()` 或 signal 取消，所以 `cancelMode` 保持 `undefined`。抛普通
`Error` 也会终止请求，并统一转换为 `ERR_INTERCEPTOR`；顶层 `message`
保留原始错误消息，原始错误保留在 `cause`，`raw` 不重复保存同一个错误。
如果后续 interceptor 的 rejected handler 返回有效配置，请求会被恢复。

### Response interceptor

```ts
const responseInterceptorId = http.interceptors.response.use(
  (response) => {
    console.log(response.statusCode)
    return response
  },
  (error) => {
    if (
      isLuchRequestError(error) &&
      error.code === LuchRequestError.ERR_NETWORK
    ) {
      console.log('网络请求失败', error.raw)
    }

    throw error
  }
)

http.interceptors.response.eject(responseInterceptorId)
```

response interceptor 抛出的错误会进入后续 rejected handler 或最终
Promise rejection。rejected handler 返回有效响应时，后续管线会恢复为
fulfilled。普通错误会统一为 `ERR_INTERCEPTOR`，并在 `response` 中保留进入
该 interceptor 的响应。

需要清空某类 interceptor：

```ts
http.interceptors.request.clear()
http.interceptors.response.clear()
```

## 错误处理

错误码通过 `LuchRequestError` 的静态常量公开。用户也可以在 interceptor
中抛出能够被请求管线识别的指定错误：

```ts
throw new LuchRequestError(
  '业务主动取消',
  LuchRequestError.ERR_CANCELED
)
```

请求管线会逐字段补充统一错误中缺失的 `config`、`task` 和 `response`；显式
传入的字段保持原值，适合高级用法。手动设置 `cancelMode` 只改变错误元数据，
不会调用原生 `Task.abort()`，普通业务不应依赖它触发取消。

| 错误码 | 说明 |
| --- | --- |
| `ERR_INVALID_CONFIG` | 配置无效，或运行环境缺少对应 uni API |
| `ERR_NETWORK` | uni API 进入 fail callback |
| `ERR_BAD_STATUS` | HTTP 状态未通过 `validateStatus` |
| `ERR_BAD_RESPONSE` | 响应内容无法按配置完成转换 |
| `ERR_CANCELED` | 调用方主动取消 |
| `ERR_INTERCEPTOR` | interceptor 执行或错误恢复失败 |

库生成的固定错误消息使用英文，便于日志检索和跨语言协作；interceptor
抛出的普通错误会保留其原始 `message`。`ERR_NETWORK` 优先使用平台 fail
错误中的非空 `errMsg`，其次使用 `message` 或字符串原因；无法提取时回退为
`Network request failed`。错误分类仍以 `error.code` 为稳定判断依据，不要匹配
`message`。调用方主动提供的错误、取消原因以及平台返回的 `raw`、`cause`
保持原值，库不会翻译。需要向最终用户展示中文等本地化提示时，应由应用根据
错误分类映射文案。

建议始终使用类型守卫：

```ts
try {
  await http.get('/users/1')
} catch (error) {
  if (!isLuchRequestError(error)) {
    throw error
  }

  console.log(error.code)
  console.log(error.config)
  console.log(error.response)
  console.log(error.task)
  console.log(error.cause)
  console.log(error.raw)
  console.log(error.cancelMode)
}
```

- HTTP 状态错误在 `response` 中保留完整响应。
- 网络失败在 `raw` 和 `cause` 中保留平台错误，并尽量将原始消息提升到 `message`。
- interceptor 普通异常在 `cause` 中保留原始错误，不重复写入 `raw`。
- 由库取消流程产生的错误通过 `cancelMode` 区分原生中断与逻辑取消。
- `isLuchRequestError()` 不依赖 `instanceof`，适合跨包和跨运行环境判断。

对于库自身产生的 `ERR_BAD_STATUS` 或 `ERR_BAD_RESPONSE`，存在
`error.response` 表示 uni API 已经返回响应。`ERR_INTERCEPTOR` 的 `response`
表示进入 interceptor 的响应上下文；如果前序 rejected handler 合成了恢复
响应，它不一定来自服务端。`ERR_NETWORK` 只能说明 uni API 报告网络失败，
不能证明服务端实际收到了请求。

`toJSON()` 返回稳定的错误摘要：

```ts
const result = error.toJSON()

result.name
result.message
result.stack
result.code
result.statusCode
result.config
result.cause
result.raw
result.cancelMode
result.isLuchRequestError
```

`statusCode` 从 `response.statusCode` 提取。原生 `task` 和完整 `response`
仍可从错误实例读取，但不会进入 `toJSON()`。`config`、`cause` 和 `raw`
保持原始引用；该方法不克隆、不脱敏，也不处理循环引用。

## 上传与下载

`upload()` 和 `download()` 返回真正的 Promise，同时附加 `abort()`、
`onTask()` 和 `task`。单次配置也可以通过 `onTask(nativeTask, control)`
在原生 Task 创建后立即注册平台事件。

### 上传

```ts
let shouldCancelUpload = false

const pendingUpload = http.upload({
  url: '/files',
  filePath: tempFilePath,
  name: 'file',
  formData: {
    category: 'avatar'
  },
  onTask(task, control) {
    task.onProgressUpdate?.((event) => {

history.md
## [3.1.1](https://github.com/lei-mu/luch-request/compare/3.1.0-alpha.0...3.1.1) (2023-08-02)




## [3.1.0](https://github.com/lei-mu/luch-request/compare/3.0.8...3.1.0) (2023-05-25)


### Features

* 更新请求新配置项。新增‘是否尝试将响应数据json化’配置项。 ([68344a1](https://github.com/lei-mu/luch-request/commit/68344a1d9dcbf586d4a0aff37d02aa0bb27ff45c)), closes [#113](https://github.com/lei-mu/luch-request/issues/113)
* 增加`version` 属性，获取插件版本号 ([40403d2](https://github.com/lei-mu/luch-request/commit/40403d2ef57dead0f577a384ac686cc7dcc772da))




## [3.0.8](https://github.com/lei-mu/luch-request/compare/3.0.7...3.0.8) (2022-08-23)


### Features

* `put` 方法增加快手小程序和京东小程序的支持 ([7e15dff](https://github.com/lei-mu/luch-request/commit/7e15dff0ac343ccbd6a298c7c9ae8ef0e2148eb9))
* 参考axio源代码 添加支持paramsSerializer。 ([cae1043](https://github.com/lei-mu/luch-request/commit/cae10430dc7222d85e0cadb13512898d66998547))
* 返回数据增加`rawData`,保留返回的原始数据 ([3924b7c](https://github.com/lei-mu/luch-request/commit/3924b7c3ea444526be436a31f9e7d5623749580b))
* 局部请求配置支持baseURL ([df2d98f](https://github.com/lei-mu/luch-request/commit/df2d98fc063f356ffd525158e9cd47d7eb24f68f))




## 3.0.7 (2021-09-04)

1. Bug Fix: 修复通过 `Request.config` 设置全局参数，多个实例`config`存在共同引用bug


## 3.0.6 (2021-05-10)

1. New Feature: APP端 增加`responseType`配置项

## 3.0.5 (2021-01-10)
### Features

* [重要] APP不再支持`CONNECT`、`HEAD`、`TRACE`请求方式。[uni.request](https://uniapp.dcloud.io/api/request/request)
* [重要]全局默认`timeout`由`30000`ms,改为`60000`ms 
* [重要]增加`index.d.ts`文件支持。感谢`Mr_Mao`的支持。github:`https://github.com/TuiMao233` 
* [重要]网络请求相关接口 uni.request、uni.uploadFile、uni.downloadFile 支持 timeout 参数。 
* [重要]返回结果response 增加`fullPath`参数。 

## 3.0.4 (2020-07-05)

1. New Feature: request 方法增加 ` firstIpv4 `配置项  
1. New Feature: 增加 ` middleware `通用请求方法

## 3.0.3 (2020-06-16)

1. Bug Fix: 修复` params ` 选项对数组格式化错误bug

## 3.0.2 (2020-06-04)

1. Bug Fix: 修复文件上传和request 配置缺少字段bug 

## 3.0.1 (2020-06-02)

1. Bug Fix: 请求方式都为` GET `的bug

## 3.0.0 (2020-06-01)

1. New Feature: 支持多拦截器
1. New Feature: 支持局部配置自定义验证器

## 2.0.1 (2020-05-01)

1. Bug Fix: 修复多实例全局配置共用问题

## 2.0.0 (2020-04-24)

1. New Feature: 增加 request ` withCredentials `选项（仅h5端支持）
1. New Feature: h5端 upload 增加 ` files ` ` file `选项。[uni.uploadFile](https://uniapp.dcloud.io/api/request/network-file?id=uploadfile "uni.uploadFile")
1. Enhancement: ` params ` 选项参数格式化方法使用axios 格式化方法
1. Bug Fix: 对upload 返回data 为空字符串的情况容错
1. Change: 修改header与全局合并方式。当前：header = Object.assign(全局，局部)

## 0.0.0 (2019-05)

1. luch-request created




request/request.js
/**
 * Request 1.0.1
 * @Class Request
 * @description luch-request 1.0.1 http请求插件
 * @Author lu-ch
 * @Date 2019-09-20
 * @Email webwork.s@qq.com
 * http://ext.dcloud.net.cn/plugin?id=392
 */
export default class Request {
  config = {
    baseUrl: '',
    header: {
      'content-type': 'application/json;charset=UTF-8'
    },
    method: 'GET',
    dataType: 'json',
    responseType: 'text'
  }

  static posUrl (url) { /* 判断url是否为绝对路径 */
    return /(http|https):\/\/([\w.]+\/?)\S*/.test(url)
  }

  static addQueryString (params) {
    let paramsData = ''
    Object.keys(params).forEach(function (key) {
      paramsData += key + '=' + params[key] + '&'
    })
    return paramsData.substring(0, paramsData.length - 1)
  }

  /**
   * @property {Function} request 请求拦截器
   * @property {Function} response 响应拦截器
   * @type {{request: Request.interceptor.request, response: Request.interceptor.response}}
   */
  interceptor = {
    /**
     * @param {Request~requestCallback} cb - 请求之前拦截,接收一个函数（config, cancel）=> {return config}。第一个参数为全局config,第二个参数为函数，调用则取消本次请求。
     */
    request: (cb) => {
      if (cb) {
        this.requestBeforeFun = cb
      }
    },
    /**
     * @param {Request~responseCallback} cb 响应拦截器，对响应数据做点什么
     * @param {Request~responseErrCallback} ecb 响应拦截器，对响应错误做点什么
     */
    response: (cb, ecb) => {
      if (cb && ecb) {
        this.requestComFun = cb
        this.requestComFail = ecb
      }
    }
  }

  requestBeforeFun (config) {
    return config
  }

  requestComFun (response) {
    return response
  }

  requestComFail (response) {
    return response
  }

  /**
   * 自定义验证器，如果返回true 则进入响应拦截器的响应成功函数(resolve)，否则进入响应拦截器的响应错误函数(reject)
   * @param { Object } response - 请求响应体（只读）
   * @return { Boolean } 如果为true,则 resolve, 否则 reject
   */
  validateStatus (response) {
    return response.statusCode === 200
  }

  /**
   * @Function
   * @param {Request~setConfigCallback} f - 设置全局默认配置
   */
  setConfig (f) {
    this.config = f(this.config)
  }

  /**
   * @Function
   * @param {Object} options - 请求配置项
   * @prop {String} options.url - 请求路径
   * @prop {Object} options.data - 请求参数
   * @prop {Object} [options.responseType = config.responseType] [text|arraybuffer] - 响应的数据类型
   * @prop {Object} [options.dataType = config.dataType] - 如果设为 json，会尝试对返回的数据做一次 JSON.parse
   * @prop {Object} [options.header = config.header] - 请求header
   * @prop {Object} [options.method = config.method] - 请求方法
   * @returns {Promise<unknown>}
   */
  async request (options = {}) {
    options.baseUrl = this.config.baseUrl
    options.dataType = options.dataType || this.config.dataType
    options.responseType = options.responseType || this.config.responseType
    options.url = options.url || ''
    options.data = options.data || {}
    options.params = options.params || {}
    options.header = options.header || this.config.header
    options.method = options.method || this.config.method
    return new Promise((resolve, reject) => {
      let next = true

      let handleRe = {}
      options.complete = (response) => {
        response.config = handleRe
        if (this.validateStatus(response)) { // 成功
          response = this.requestComFun(response)
          resolve(response)
        } else {
          response = this.requestComFail(response)
          reject(response)
        }
      }
      const cancel = (t = 'handle cancel', config = options) => {
        const err = {
          errMsg: t,
          config: config
        }
        reject(err)
        next = false
      }

      handleRe = { ...this.requestBeforeFun(options, cancel) }
      const _config = { ...handleRe }
      if (!next) return

      let mergeUrl = Request.posUrl(options.url) ? options.url : (options.baseUrl + options.url)
      if (JSON.stringify(options.params) !== '{}') {
        const paramsH = Request.addQueryString(options.params)
        mergeUrl += mergeUrl.indexOf('?') === -1 ? `?${paramsH}` : `&${paramsH}`
      }
      _config.url = mergeUrl
      uni.request(_config)
    })
  }

  get (url, options = {}) {
    return this.request({
      url,
      method: 'GET',
      ...options
    })
  }

  post (url, data, options = {}) {
    return this.request({
      url,
      data,
      method: 'POST',
      ...options
    })
  }

  // #ifndef MP-ALIPAY
  put (url, data, options = {}) {
    return this.request({
      url,
      data,
      method: 'PUT',
      ...options
    })
  }

  // #endif

  // #ifdef APP-PLUS || H5 || MP-WEIXIN || MP-BAIDU
  delete (url, data, options = {}) {
    return this.request({
      url,
      data,
      method: 'DELETE',
      ...options
    })
  }

  // #endif

  // #ifdef APP-PLUS || H5 || MP-WEIXIN
  connect (url, data, options = {}) {
    return this.request({
      url,
      data,
      method: 'CONNECT',
      ...options
    })
  }

  // #endif

  // #ifdef APP-PLUS || H5 || MP-WEIXIN || MP-BAIDU
  head (url, data, options = {}) {
    return this.request({
      url,
      data,
      method: 'HEAD',
      ...options
    })
  }

  // #endif

  // #ifdef APP-PLUS || H5 || MP-WEIXIN || MP-BAIDU
  options (url, data, options = {}) {
    return this.request({
      url,
      data,
      method: 'OPTIONS',
      ...options
    })
  }

  // #endif

  // #ifdef APP-PLUS || H5 || MP-WEIXIN
  trace (url, data, options = {}) {
    return this.request({
      url,
      data,
      method: 'TRACE',
      ...options
    })
  }

  // #endif

  upload (url, {
    // #ifdef APP-PLUS
    files,
    // #endif
    // #ifdef MP-ALIPAY
    fileType,
    // #endif
    filePath,
    name,
    header,
    formData
  }) {
    return new Promise((resolve, reject) => {
      let next = true
      let handleRe = {}
      const pubConfig = {
        baseUrl: this.config.baseUrl,
        url,
        // #ifdef APP-PLUS
        files,
        // #endif
        // #ifdef MP-ALIPAY
        fileType,
        // #endif
        filePath,
        method: 'UPLOAD',
        name,
        header: header || this.config.header,
        formData,
        complete: (response) => {
          response.config = handleRe
          if (response.statusCode === 200) { // 成功
            response = this.requestComFun(response)
            resolve(response)
          } else {
            response = this.requestComFail(response)
            reject(response)
          }
        }
      }
      const cancel = (t = 'handle cancel', config = pubConfig) => {
        const err = {
          errMsg: t,
          config: config
        }
        reject(err)
        next = false
      }

      handleRe = { ...this.requestBeforeFun(pubConfig, cancel) }
      const _config = { ...handleRe }
      if (!next) return
      _config.url = Request.posUrl(url) ? url : (this.config.baseUrl + url)
      uni.uploadFile(_config)
    })
  }
}

/**
 * setConfig回调
 * @return {Object} - 返回操作后的config
 * @callback Request~setConfigCallback
 * @param {Object} config - 全局默认config
 */
/**
 * 请求拦截器回调
 * @return {Object} - 返回操作后的config
 * @callback Request~requestCallback
 * @param {Object} config - 全局config
 * @param {Function} [cancel] - 取消请求钩子，调用会取消本次请求
 */
/**
 * 响应拦截器回调
 * @return {Object} - 返回操作后的response
 * @callback Request~responseCallback
 * @param {Object} response - 请求结果 response
 */
/**
 * 响应错误拦截器回调
 * @return {Object} - 返回操作后的response
 * @callback Request~responseErrCallback
 * @param {Object} response - 请求结果 response
 */
